### PR TITLE
feature/985 branch protection `checks`

### LINKED ISSUE #985: Add support for configuring checks to branch protection
### Affected Resource(s)
Please list the resources as a list, for example:
- github_branch_protection
- github_branch_protection_v3

### Description

The [GitHub API](https://docs.github.com/en/rest/reference/repos#update-branch-protection) supports setting up `checks` as well as `contexts` for branch protection. `checks` provide a few more configuration options (like enforcing that a check is set by a specific app)

image

Could you add support to the `github_branch_protection` and `github_branch_protection_v3` resources to configure `checks`?  Currently they only support `contexts`

Thanks

### PR DESCRIPTION
**Pull request by TheQueenIsDead**
Resolves: 
 - #985
 - #1147 
 - #1307
 - #1212

----

### Before the change?

* Checks were not able to be set 
* Setting required status checks on branch protection resources was broken

### After the change?

* Checks are able to be set via a ":" separated string
* Contexts are now passed to the API as checks and users can provision branch protections again

### Other information

This should render the ability to define checks in configuration such as this:
```tf
resource "github_branch_protection_v3" "bp" {

  repository = "terraform-template-module"
  branch     = "master"

  required_status_checks  {
    strict = true

    contexts = [
      "context1",
      "context2",
      "context3",
      "context4",
    ]

    checks = [
      "negative_one:-1",
      "undefined_comma:",
      "undefined",
      "zero_check:0",
      "app_id_context:1234567",
    ]
  }
}
```

----

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:  

* N/A

### Pull request type

 
 

Please add the corresponding label for change this PR introduces:
- Bugfix: `Type: Bug`
- Feature/model/API additions: `Type: Feature`
- Updates to docs or samples: `Type: Documentation`
- Dependencies/code cleanup: `Type: Maintenance`

----

### CODE REVIEW

**Comment by TheQueenIsDead** [comment]
This works in the sense that is creates and deletes branch protections :+1: 

However, there is something interesting happening in the diff, like so:
```
 # github_branch_protection_v3.bp will be updated in-place
  ~ resource "github_branch_protection_v3" "bp" {
        id                              = "terraform-template-module:master"
        # (6 unchanged attributes hidden)

      ~ required_status_checks {
          ~ contexts       = [
              - "example",
              - "newcheck",
              - "noappidcontext",
            ]
            # (2 unchanged attributes hidden)

          ~ check {
              ~ app_id  = -1 -> 1
                # (1 unchanged attribute hidden)
            }
          - check {
              - app_id  = -1 -> null
              - context = "newcheck" -> null
            }

            # (1 unchanged block hidden)
        }
    }
```

This seems to suggest that despite not including a `contexts` array, they will become updated via the inclusion of `check` blocks.

Not too sure how to handle this currently, but will see what I can come up with.

File: github/resource_github_branch_protection_v3_utils.go [hunk 1]
```diff
@@ -46,10 +46,24 @@ func flattenAndSetRequiredStatusChecks(d *schema.ResourceData, protection *githu
 			contexts = append(contexts, c)
 		}
 
+		var checks []interface{}
+		for _, chk := range rsc.Checks {
+			chkMap := make(map[string]interface{})
+			chkMap["context"] = chk.Context
+			chkMap["app_id"] = func() int {
+				if chk.AppID != nil {
```
Comment (TheQueenIsDead) [inline github/resource_github_branch_protection_v3_utils.go:54]: If this is the ID from upstream via the SDK, it's likely to already be defaulted to a non-nil making this check redundant.

File: github/resource_github_branch_protection_v3.go [hunk 2]
```diff
@@ -53,12 +53,30 @@ func resourceGithubBranchProtectionV3() *schema.Resource {
...
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"context": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"app_id": {
+										Type:     schema.TypeInt,
+										Optional: true,
+										Default:  -1,
```
Comment (TheQueenIsDead) [inline github/resource_github_branch_protection_v3.go:75]: Upon reflection this should be nullable.
The API states the field can be _ommitted_ to allow the most recent app, **or** set to _-1_ to allow any app:
https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection
https://pkg.go.dev/github.com/google/go-github/v48@v48.1.0/github#RequiredStatusCheck

**Comment by cfisher281** [comment]
Appreciate you doing this work @TheQueenIsDead.

**Comment by TheQueenIsDead** [comment]
@kfcampbell , small question from me. I've added the deprecated flag to the contexts attribute, and the process for carrying out this deprecation is usually to release a minor with the flag, and then remove the code in the next major release version. This is documented by Hashicorp here:
https://developer.hashicorp.com/terraform/plugin/sdkv2/best-practices/deprecations

It makes a lot of sense, but given the change to the Github go client, I cannot implement this functionality nicely without removing part of the code that flattens the context array when it reads from the API. Without that removal, subsequent applies will show the contexts array being blitzed on each plan, despite it not being defined in Terraform:
```console
 ~ contexts       = [
              - "context_app_id_allow_all",
              - "context_no_app_id_most_recent_check",
              - "context_with_app_id_defined",
            ]
```

The question is, would you accept the removal of `contexts` in this PR (Jarring as on release it would give little time to warn users of deprecation), or is there perhaps another route to releasing this that would be preferred?

As it stands, it is not possible to only send `contexts` with the SDK :-( So these are interesting circumstances.

**Comment by kfcampbell** [comment]
Hmm...none of those options are very desirable. 

I'm looking at the [API reference](https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection) and I see the following under `required_status_checks`: 

image

> Deprecated: The list of status checks to require in order to merge into this branch. If any of these checks have recently been set by a particular GitHub App, they will be required to come from that app in future for the branch to merge. Use checks instead of contexts for more fine-grained control.

Perhaps some of that behavior is due to the API changing to deprecate contexts as well. 

> The question is, would you accept the removal of contexts in this PR (Jarring as on release it would give little time to warn users of deprecation), or is there perhaps another route to releasing this that would be preferred?

For a breaking change like that, we would end up cutting a v6 version, which ideally we should not do that in this PR without a deprecation notice. 

Do you mind explaining a little bit more about this description? 

> given the change to the Github go client, I cannot implement this functionality nicely without removing part of the code that flattens the context array when it reads from the API. Without that removal, subsequent applies will show the contexts array being blitzed on each plan, despite it not being defined in Terraform:

Is there a way we can handle contexts when they're present in Terraform by warning of deprecation and then formatting them into the new `checks` array of objects?

**Comment by TheQueenIsDead** [comment]
> Perhaps some of that behavior is due to the API changing to deprecate contexts as well.

I wouldn't be suprised! As identified in a previous issue, the SDK "fully deprecated" contexts as part of this PR, while I believe the API itself still supports it momentarily:
https://github.com/google/go-github/issues/2467#issuecomment-1250070578

> For a breaking change like that, we would end up cutting a v6 version, which ideally we should not do that in this PR without a deprecation notice.

I appreciate that, and am happy to help go down that route where possible 👍 However, it does feel a little ceremonious given `contexts` on the SDK are broken due to the above PR.

> Do you mind explaining a little bit more about this description?

Certainly! What I am trying to highlight is that the Github API seems to return a list of `contexts` even when only `checks` are provisioned. When this gets converted to TF schema, it assumes that on next apply the `contexts` array will be deleted, because it is not defined in our file.

I think you're right that we could simply convert contexts to checks (With a default of allow all apps, ie -1). We could possibly keep the schema and not try and map the response to a `contexts` array in that instance.
That feels a little dirty, but it could be a good midpoint to allowing users to keep doing what they're doing, and give them a little notice about moving to checks.

I'll have a play around :-)

**Comment by TheQueenIsDead** [comment]
Hmm, super interesting. I have made a change to flatten and expand all contexts into checks, using the default -1 app_id.

On initial apply, everything looks ok. When a second apply is ran, the change output looks really messy. On the third apply in a row (With no changes), Terraform detects no changes. 

I would have expected it to detect no changes the first time around, but I am unsure why it creates a plan like this. Full example output can be seen in this gist:
https://gist.github.com/TheQueenIsDead/ea838b56322f4aa998cd1b84f1129cbe

**Comment by TheQueenIsDead** [comment]
I have tried sorting the `contexts` and `checks` array alphabetically, but this did not resolve the weird diff. I feel that this may be due to the way Terraform does the diff (Which I am not too familiar with under the hood), but I am guessing it reconciles diffs based on it's own schema, where it has one schema based on current state, and another based on what the file definition says.

I also tried experimenting with `ConfigMode: schema.SchemaConfigModeAttr`, which did not seem to help.

I'm wondering if the 'advanced data structure' as an array is causing issues, I imagine if we used a map of strings `([]string{"context_name:-1"})` then we would get a better diff, though this is inflexible for representing new attributes that may be added to the API in the future.

**Comment by TheQueenIsDead** [comment]
@kfcampbell , I have changed the approach from using resource blocks in the schema to instead using a list of `:` delimited strings. This yielded a much better diff, and has the added bonus that an end user can simply rename the `contexts` array and have a working config.

Will update the docs soon, not terribly sure about how to approach testing.

**Comment by kfcampbell** [comment]
> I have changed the approach from using resource blocks in the schema to instead using a list of : delimited strings. This yielded a much better diff, and has the added bonus that an end user can simply rename the contexts array and have a working config.

:+1:, thank you for the persistence on this one. 

> Will update the docs soon

Docs changes look good. As a side note, since the current behavior is broken as you mentioned, that breaking change feels okay to release under v5.x.x.

> not terribly sure about how to approach testing.

Our testing in the project isn't in a very healthy state at the moment. While #1414 and other related issues are still open, I've been running integration tests manually before merging PRs :see_no_evil: :see_no_evil: :see_no_evil: 

You can find some documentation and an example test pattern to follow [here](https://github.com/integrations/terraform-provider-github/blob/main/CONTRIBUTING.md#local-development-iteration) in our CONTRIBUTING.md file. Please let me know if you have more specific questions!

**Comment by TheQueenIsDead** [comment]
No worries! Thanks for bearing with my first time developing a provider.

> since the current behavior is broken as you mentioned, that breaking change feels okay to release under v5.x.x.

If we go with the list approach then we should still be able to use contexts and checks, meaning the change will not be breaking :-) 

Alrighty, so I created a new org and setup the tests on my forked repo's main. All of the tests pass except for the `required_status_check` test which fails with the expected schema violations caused by the [v48.1.0](https://github.com/TheQueenIsDead/terraform-provider-github/blob/0b3d26d9a28509eec7b4f1df37f0a57ca798260a/go.mod#L8) version of `go-github`
image


However, this is the `main` branch of **MY** fork, which is 13 commits behind the official repositories main:
https://github.com/TheQueenIsDead/terraform-provider-github/tree/main

When I run the tests on this PR branch (which is up to date with the latest version of the `integrations` organisation main branch) we're not as successful :see_no_evil: Lots each test for branch protections v3 fails on:
```
Error: json: cannot unmarshal object into Go struct field Protection.lock_branch of type bool
``` 

I wonder if this breakage was introduced by the recent update to the go-github provider version with support for `lock_branch` as introduced by this PR https://github.com/google/go-github/pull/2577 in [v48.2.0](https://github.com/google/go-github/releases/tag/v48.2.0)

I see that this was updated recently 
https://github.com/integrations/terraform-provider-github/pull/1425

@kfcampbell is there any chance that this has broken the integration tests on main for the `GithubBranchProtectionV3` tests?

**Comment by TheQueenIsDead** [comment]
I have just synced my fork with master and I can confirm that the tests are not passing on main anymore. @kfcampbell does this require a issue to be raised? It slightly blocks me on testing unless I revert the version back

**Comment by fproulx-boostsecurity** [comment]
Yay

**Comment by gilacost** [comment]
Any updates on this? I was looking around and trying to contribute but I found out that this was already in place. It looks great! What does else do we need to do to get it merged? 
Please add also this [issue](https://github.com/integrations/terraform-provider-github/issues/1451) to the closing block in the PR description. 🤗

**Comment by kfcampbell** [comment]
@TheQueenIsDead thank you for all your effort and clear communication! I really appreciate it. 

> is there any chance that this has broken the integration tests on main for the GithubBranchProtectionV3 tests?

Definitely, and there's also a chance that they haven't been exactly healthy for a long time. 

> does this require a issue to be raised?

Ideally, yes. We have a significant amount of testing debt to pay back on this project. I've opened #1414 to begin this effort, but it would be great to have an issue for this specific testing suite. 

Given the above and the substantial number of issues ready to be closed, I think it would be reasonable to merge this PR before full testing support is achieved.

@TheQueenIsDead do you believe this is ready for final review/merging?

File: github/resource_github_branch_protection_v3_utils.go [hunk 3]
```diff
@@ -40,16 +40,34 @@ func buildProtectionRequest(d *schema.ResourceData) (*github.ProtectionRequest,
...
+			// Parse into contexts
 			contexts = append(contexts, c)
+			checks = append(contexts, c)
+		}
+
+		// Flatten checks
+		for _, chk := range rsc.Checks {
+			// Parse into contexts
+			contexts = append(contexts, chk.Context)
+			checks = append(contexts, fmt.Sprintf("%s:%d", chk.Context, chk.AppID)) //  buildTwoPartID(chk.Context, strconv.Itoa(int(*chk.AppID))))
```
Comment (kfcampbell) [inline github/resource_github_branch_protection_v3_utils.go:62]: Can this end-of-line comment be removed?

File: github/resource_github_branch_protection_v3_utils.go [hunk 4]
```diff
@@ -191,8 +209,56 @@ func expandRequiredStatusChecks(d *schema.ResourceData) (*github.RequiredStatusC
 			m := v.(map[string]interface{})
 			rsc.Strict = m["strict"].(bool)
 
+			// Initialise empty literal to ensure an empty array is passed mitigating schema errors like so:
+			// For 'anyOf/1', {"strict"=>true, "checks"=>nil} is not a null. []
+			rscChecks := []*github.RequiredStatusCheck{}
+
+			// TODO: Remove once contexts is deprecated
+			// Iterate and parse contexts into checks using -1 as default to allow checks from all apps.
```
Comment (kfcampbell) [inline github/resource_github_branch_protection_v3_utils.go:217]: Is this -1 value documented on the GitHub API? Do you know where it comes from?

**Review by kfcampbell — COMMENTED** [review]
Hi! I'm swinging by to check on this PR again as I'm really looking forward to having it in. @TheQueenIsDead do you mind giving these questions a look when you get a chance?

File: github/resource_github_branch_protection_v3_utils.go (hunk 4 above)
Comment (TheQueenIsDead) [reply github/resource_github_branch_protection_v3_utils.go:217]: Yes it is :-) 
https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection
image

**Comment by TheQueenIsDead** [comment]
Ok, so I have created a new bug report for the unmarshalling error that is causing the tests to fail for this branch #1476.

> Given the above and the substantial number of issues ready to be closed, I think it would be reasonable to merge this PR before full testing support is achieved. @TheQueenIsDead do you believe this is ready for final review/merging?

In a perfect world I think the recent go-github provider breakage would be reverted so that the tests can pass. As mentioned in the issue, it seems to have broken unmarshalling branch protection objects, so even if this PR is merged and released, users still wont be able to benefit from it!

I'm happy with the changes proposed here, but I am also the author, hence the most biased party ;-) The lack of tests are the only thing making me wary however.
The diff was a little confusing when running an apply to get the contexts to be written as checks, but subsequent runs yeilded a decent diff, which I am happy about given my limited experience with programming and the Terraform SDK

**Comment by kfcampbell** [comment]
> I have created a new bug report for the unmarshalling error that is causing the tests to fail for this branch https://github.com/integrations/terraform-provider-github/issues/1476.

This should be fixed. I believe the problem was due to [this upstream issue](https://github.com/google/go-github/issues/2612), fixed in [this google/go-github PR](https://github.com/google/go-github/pull/2606), released [here](https://github.com/google/go-github/releases/tag/v49.0.0), and added to the project as part of #1473. 

On the main branch, tests for this resource are failing for me as well :sob: but not for unmarshaling reasons. 

> it seems to have broken unmarshalling branch protection objects, so even if this PR is merged and released, users still wont be able to benefit from it!

I don't believe this is the case anymore! :tada: 

> The lack of tests are the only thing making me wary

Absolutely, I agree. However, given the resource's impaired state currently and the number of issues/engagement about it, it may be appropriate to merge this PR earlier as a mitigation and add/modify earlier content afterward. Issues saying "confusing diff" are generally better than issues saying "doesn't work" IMO.

**Comment by ilmax** [comment]
I also bumped into this one, looking forward to seeing this merged :)

**Comment by TheQueenIsDead** [comment]
Apologies for the delay, I've been struggling to find time to commit to other projects.

> On the main branch, tests for this resource are failing for me as well sob but not for marshalling reasons.

I'll have a quick look at the tests to see if I can identify the issue and resolve it here, otherwise I'm all for merging this to give a path forwards, with room to refine later on. Pragmatism 😊

**Comment by kfcampbell** [comment]
@TheQueenIsDead thank you for all the work here! 

I realized I merged #1489; does that affect this PR/testing scenarios at all? If you're comfortable with the current state after #1489, I'll approve/merge this and cut a release soon.

**Comment by TheQueenIsDead** [comment]
Awesome, that resolves all the issues that aren't to do with `required_status_checks`!
The only issue with this PR is that it takes one more apply before it is happy reconciling the diffs on future runs.

I'm happy to have this merged with the goal of improving it later :+1:

**Review by kfcampbell — APPROVED** [review]
Thanks for all the work on this, @TheQueenIsDead! I'm looking forward to having this in.

**Comment by rpdelaney** [comment]
Any guess when we might get this in a release? :)

**Comment by mbarany** [comment]
what if my status check context includes a colon?
Our CircleCi jobs for example `ci/circleci: unit_test` produces the following error:
```
Error: Could not parse unit_test:-1 as valid app_id
```

**Comment by kfcampbell** [comment]
@mbarany do you mind creating an issue for that?

**Pull request merged**